service/tablet_allocator.cc

```diff
@@ -877,6 +879,158 @@ class load_balancer {
+                    if (r1_it != r1_map.end()) {
+                        return std::make_pair(r1_it->second, r2[i]);
+                    }
+                }
+                // Since sets had intersection subtracted, the remaining replicas are certainly not co-located.
+                if (r1.size() > 0) {
+                    return std::make_pair(r1[0], r2[0]);
+                }
+                return std::nullopt;
+            };
+
+            auto create_migration_info = [] (global_tablet_id gid, tablet_replica src, tablet_replica dst) {
+                auto kind = (src.host != dst.host) ? tablet_transition_kind::migration : tablet_transition_kind::intranode_migration;
+                return tablet_migration_info{kind, gid, src, dst};
+            };
+
+            bool all_colocated = true;
+            co_await tmap.for_each_sibling_tablets([&] (tablet_desc t1, std::optional<tablet_desc> t2_opt) -> future<> {
+                // Be optimistic about migrating tablets, as if they succeeded.
+                // Merge finalization will have to recheck that all sibling tablets are co-located.
+
+                if (!t2_opt) {
+                    on_internal_error(lblogger, format("Unable to find sibling tablet during co-location, with tablet count {}, for table {}",
+                                                       tmap.tablet_count(), table));
+                }
+                auto t2 = *t2_opt;
+
+                auto r1 = get_replicas(t1);
+                auto r2 = get_replicas(t2);
+                if (r1 == r2) {
+                    return make_ready_future<>();
+                }
+                all_colocated = false;
+
+                if (migrating(t1) || migrating(t2)) {
+                    return make_ready_future<>();
+                }
+                // During RF change, tablets may have incrementally replicas allocated / deallocated to them.
+                // Let's temporarily delay their co-location until their replica sets have the same size.
+                if (r1.size() != r2.size()) {
+                    lblogger.warn("Replica sets of tablets to be co-located differ in size: ({}: {}), ({}, {})",
+                                  t1.tid, r1, t2.tid, r2);
+                    return make_ready_future<>();
+                }
+
+                // Returns true if moving candidate into dst will violate replication constraint.
+                const auto r2_hosts = r2
+                    | std::views::transform(std::mem_fn(&locator::tablet_replica::host))
+                    | std::ranges::to<std::unordered_set<host_id>>();
+                auto check_constraints = [r2_hosts = std::move(r2_hosts)] (tablet_replica src, tablet_replica dst) {
+                    // handles intra-node migration.
+                    if (src.host == dst.host && src.shard != dst.shard) {
+                        return false;
+                    }
+                    return r2_hosts.contains(dst.host);
```

Comment:
This doesn't check for rack overload, which means we may temporarily overload racks until all is colocated. We can probably live with it for now. @bhalevy

Reply:
We can possibly improve it by sorting replica set by rack id. So with RF=#_of_racks, co-location should not increase  load on any rack.

Reply:
Postponing this since it requires some more careful thought.